Document results from team discussions for eventing team code guidelines (#634)
kyma-project/community: concepts/eventing-code-guidelines.md

**Assertion libraries**:

> Use [stretchr/testify](https://github.com/stretchr/testify) as assertion library. The only exception to this are the controller integration tests that use [onsi/gomega](https://github.com/onsi/gomega).

Reviewer: cant we say something like: if in doubt use strechr, unless there are valid reasons not to use it?

Author: What would be the benefit 🤔 ? To make it not required but only recommended ? If yes then I think this can go into the [section header]( https://github.com/kyma-project/community/blob/2f86f9d443d1906ec8c1e0c7c166f42b2180c8a9/concepts/eventing-code-guidelines.md#recommended-libraries). The section header already talks about **desired choice**.

> The following section describes the desired choice of libraries for testing, logging, etc.

Author: @k15r What about this? The word `only` might be the problem in the above sentence:

```suggestion
Use [stretchr/testify](https://github.com/stretchr/testify) as assertion library. The controller integration tests are an exception to this and use [onsi/gomega](https://github.com/onsi/gomega) as an assertion library instead.
```

Author: @raypinto So far they are the only exception. We migrated the controller integration tests for the nats controller from `ginkgo` to `testing.T` (see this [PR](https://github.com/kyma-project/kyma/pull/13409)), but gomega is still the assertion library (also true for beb and backend reconcilers).
We did not decide as an eventing team whether we want to go away from gomega for the controller integration tests so far. Unit tests however shall not use gomega at all. This is why controller integration tests are an exception for now (can change of course).

Reviewer 2: @nachtmaar From our last discussion in the team, I understood that the only reason we need gomega assertions is because we thought that do not have "eventually" in strectr/testify library. Now that we know that it exists, do we still need this exception? Can we update the existing tests to use strectr/testify library completely?

Author: @raypinto Totally correct 👍 IMHO we should do a POC first to see how the controller integration tests look then being rewritten to use `stretchr/testify` as assertion library. Until we have this POC (and if the team prefers it), we can have this exception.
@raypinto  I can create a ticket for the POC. What do you think ?

Reviewer 2: @nachtmaar Sounds good!

Author: @raypinto I created an EPIC for further code migration and made the POC one subtask of it:
- EPIC: https://github.com/kyma-project/kyma/issues/13559
- POC: https://github.com/kyma-project/kyma/issues/13565

> Use [stretchr/testify](https://github.com/stretchr/testify) as assertion library. The controller integration tests are an exception to this and use [onsi/gomega](https://github.com/onsi/gomega) as an assertion library instead.

Reviewer 3: (nit)
```suggestion
Use [stretchr/testify](https://github.com/stretchr/testify) as assertion library. The controller integration tests are an exception to this, use [onsi/gomega](https://github.com/onsi/gomega) as an assertion library for them instead.
```

for me it sounds better

Author: I am ok with both ways. @NHingerl what do you prefer ?

Reviewer 4: Either the version with the "and", or use a semicolon or full stop instead of the comma to separate the sentences, as they have different subjects.

Author: @VladislavPaskar I agree with @NHingerl. I think the subject is not important in this case (whether we or the tests use the assertion library because the outcome is the same).
Can you create an updated suggestion ?

Author: @VladislavPaskar As discussed: I will pick your suggestion since the sentence is more clear this way to you.
However I prefer a full stop instead of a comma.

```suggestion
Use [stretchr/testify](https://github.com/stretchr/testify) as assertion library. The controller integration tests are an exception to this. Use [onsi/gomega](https://github.com/onsi/gomega) as an assertion library instead.
```
